dali/python/nvidia/dali/plugin/paddle.py

```diff
@@ -307,30 +303,20 @@ def __next__(self):
                 else:
                     category_place[cat] = pd_cpu_place
 
-            if self._data_batches[i] is None:
-                pd_tensors = {}
-                for cat, lod in self.normalized_map.items():
-                    lod_tensor = fluid.core.LoDTensor()
-                    lod_tensor._set_dims(category_shapes[cat])
-                    pd_tensors[cat] = lod_tensor
-                self._data_batches[i] = pd_tensors
-            else:
-                pd_tensors = self._data_batches[i]
-
-            # Copy data from DALI Tensors to LoDTensors
+            pd_tensors = {}
             for cat, tensor in category_tensors.items():
-                if hasattr(tensor, 'shape'):  # could be tensor list
-                    assert self._dynamic_shape or \
-                        tensor.shape() == pd_tensors[cat].shape(), \
-                        ("Shapes do not match: DALI tensor has size {0}, "
-                         "but LoDTensor has size {1}".format(
-                             tensor.shape(), pd_tensors[cat].shape()))
-
-                lod_tensor = pd_tensors[cat]
+                lod_tensor = fluid.core.LoDTensor()
+                pd_tensors[cat] = lod_tensor
                 lod_tensor._set_dims(category_shapes[cat])
                 seq_len = category_lengths[cat]
                 lod_tensor.set_recursive_sequence_lengths(seq_len)
-                ptr = lod_tensor._mutable_data(category_place[cat],
+                lod_tensor._mutable_data(category_place[cat],
+                                         category_pd_type[cat])
+            data_batches[i] = pd_tensors
+
+            fluid.core._cuda_synchronize(pd_gpu_place)
```

Comment:
:thinking: 
Why do we have to sync?

Reply:
Because https://github.com/PaddlePaddle/Paddle/issues/35555 (at least that is the only way I managed to get it working).